src/actions/measurements.js

```diff
@@ -0,0 +1,121 @@
+ * If the current `measurementsGrouping` does not exist in the collection, then
+ * defaults to the first grouping option.
+ * @param {Object} collection
+ * @returns {Object}
+ */
+const getCollectionDisplayControls = (controls, collection) => {
+  const measurementsControls = [
+    "measurementsGroupBy",
+    "measurementsDisplay",
+    "measurementsShowOverallMean",
+    "measurementsShowThreshold"
+  ];
+  // Copy current control options for measurements
+  const newControls = pick(controls, measurementsControls);
+  // Checks the current group by is available as a field in collection
+  if (!collection.fields.has(newControls.measurementsGroupBy)) {
+    // If current group by is not available as a field, then default to the first grouping option.
+    newControls.measurementsGroupBy = collection.groupings[0];
+  }
+
+  if (collection["display_defaults"]) {
+    const {
+      group_by,
+      measurement_display,
+      show_overall_mean,
+      show_threshold
+    } = collection["display_defaults"];
+
+    if (group_by) {
+      newControls.measurementsGroupBy = group_by;
+    }
+    if (measurement_display) {
+      newControls.measurementsDisplay = measurement_display;
+    }
+    if (typeof show_overall_mean === "boolean") {
+      newControls.measurementsShowOverallMean = show_overall_mean;
+    }
+    if (typeof show_threshold === "boolean") {
+      newControls.measurementsShowThreshold = show_threshold;
+    }
+  }
+
+  return newControls;
+};
+
+export const loadMeasurements = (json) => (dispatch, getState) => {
+  // TODO: Load controls from state to get potential url query parameters
+  const { tree, controls } = getState();
+  if (!tree.loaded) {
+    throw new Error("tree not loaded");
+  }
+
+  const collections = json["collections"];
+  if (!collections || collections.length === 0) {
+    throw new Error("Measurements JSON does not have collections");
+  }
+
+  collections.forEach((collection) => {
+    // Convert provided fields array to Map for easier access and to keep ordering
+    const collectionFieldsArray = collection.fields;
+    collection.fields = new Map();
+    if (collectionFieldsArray && collectionFieldsArray.length > 0) {
+      collectionFieldsArray.forEach(({key, title}) => collection.fields.set(key, {title: title || key}));
+    }
+
+    // Get all fields from first measurment to include them if they are not in collection.fields
+    Object.keys(collection.measurements[0]).forEach((fieldKey) => {
+      if (!collection.fields.has(fieldKey)) {
+        collection.fields.set(fieldKey, {title: fieldKey});
+      }
+    });
+
+    // Add jitter and stable id for each measurement
+    collection.measurements.forEach((measurement, index) => {
+      // Generate Gaussian jitter with a Box-Muller transform
+      measurement["measurementJitter"] = Math.sqrt(-2*Math.log(Math.random()))*Math.cos(2*Math.PI*Math.random());
+      measurement["measurementId"] = index;
```

Comment:
Should add these two reserved keys to the schema (in the Augur repo) as forbidden properties. Alternatively, prefix these here with a reserved char (e.g. `__` to form `__measurementId`) and then forbid any properties starting with that (e.g. `^(?!__).+$`) in the schema.

Reply:
or use symbols ;)

Reply:
I had not made these reserved properties since I was not sure if their values should be generated within auspice (i.e. should they be generated by augur), but feedback from the Nextstrain meeting made it clear these values should be within auspice only since they are only relevant for the visualization. 

I'll change them to symbols (gotta use the new shiny thing I just learned about 😉 )

Reply:
Symbols are a much better solution!

Reply:
Resolved in https://github.com/nextstrain/auspice/pull/1452/commits/a2d414d67c60e3c469355816cff5761507d02e8d